src/scatter.jl

```diff
@@ -0,0 +1,189 @@
+export scatter!, scatter
+
+## Scatter API
+#   - Scatter:
+#     - scatter(op, src, idx)
+#     - scatter!(op, dst, src, idx)
+#   - Scatter destination backpropagation
+#     - ∇scatter_dst!
+#   - Scatter source backpropagation
+#     - ∇scatter_src
+#     - ∇scatter_src!
+#
+
+function _check_dims(Ndst, Nsrc, N, Nidx)
+    @assert Ndst - N == Nsrc - Nidx
+    dims = Ndst - N
+    if dims < 0
+        throw(ArgumentError("dims must be non-negative but got dims=$dims"))
+    end
+    return dims
+end
+
+_check_input(idx::AbstractArray{<:Integer}, arr) = checkbounds(arr, minimum(idx):maximum(idx))
+
+function _check_input(idx::AbstractArray{<:Tuple}, arr)
+    pairs = map(xs -> Base.OneTo(maximum(xs)), zip(idx...))
+    checkbounds(arr, pairs...)
+end
+
+function _check_output(idx::AbstractArray{<:IntOrTuple}, arr, src, dims)
+    pre_dims = axes(src)[1:dims]
+    post_dims = Base.OneTo.(maximum_dims(idx))
+    checkbounds(arr, pre_dims..., post_dims...)
+end
+
+
+"""
+    scatter!(op, dst, src, idx)
+
+Scatter operation, which scatters data in `src` and assigns to `dst` according to `idx`.
+With the data going to the same place, specified operation is applied on to reduce data.
+For each index `k` in `idx`, accumulate values in `dst` according to
+
+    dst[idx[k]...] = (op).(dst[idx[k]...], src[k...])
+
+# Arguments
+- `op`: operations to be applied on `dst` and `src`, e.g. `+`, `-`, `*`, `/`, `max` and `min`.
+- `dst`: the destination for `src` to aggregate to. This argument will be mutated.
+- `src`: the source data for aggregating.
+- `idx`: the mapping for aggregation from source (index) to destination (value).
+The index of `idx` is corresponding to the index of `src` and the value of `idx` is
+corresponding to the index of `dst`. The value of `idx` can be `Int` or `Tuple` type.
+
+    dst[:, ..., idx[k]...] = (op).(dst[:, ..., idx[k]...], src[:, ..., k...])
+"""
+function scatter!(op,
+                  dst::AbstractArray{Tdst,Ndst},
+                  src::AbstractArray{Tsrc,Nsrc},
+                  idx::AbstractArray{<:Integer,Nidx}) where {Tdst<:Real,Tsrc<:Real,Ndst,Nsrc,Nidx}
+    dims = _check_dims(Ndst, Nsrc, 1, Nidx)
+    @boundscheck _check_output(idx, dst, src, dims)
+    @boundscheck _check_input(idx, src)
+    T = promote_type(Tdst, Tsrc)
+    scatter!(op, T.(dst), src, idx, Val(dims))
+end
+
+function scatter!(op,
+                  dst::AbstractArray{Tdst,Ndst},
+                  src::AbstractArray{Tsrc,Nsrc},
+                  idx::AbstractArray{NTuple{N,Int},Nidx}) where {Tdst<:Real,Tsrc<:Real,Ndst,Nsrc,N,Nidx}
+    dims = _check_dims(Ndst, Nsrc, N, Nidx)
+    @boundscheck _check_output(idx, dst, src, dims)
+    @boundscheck _check_input(idx, src)
+    T = promote_type(Tdst, Tsrc)
+    scatter!(op, T.(dst), src, idx, Val(dims))
+end
+
+function scatter!(op, dst::AbstractArray{Tdst}, src::AbstractArray{Tsrc}, idx::AbstractArray{<:IntOrTuple},
+                  dims::Val{N}) where {Tdst<:Real,Tsrc<:Real,N}
```

Comment:
```suggestion
function scatter!(op, dst::AbstractArray{Tdst}, src::AbstractArray{Tsrc}, idx::AbstractArray{<:IntOrTuple},
                  dims::Val{N}) where {Tdst,Tsrc,N}
```
Also, probably we don't need an extra method with the additional `dims` argument. `dims` can be calculated inside. In #280 that didn't seem to have  a performance impact, but maybe you can benchmark it by yourself here.

Reply:
It may provide pre-compilation information for compiler to compile specialized function for `dims` and dispatch on it. Constant propagation helps a lot.

Reply:
I think the compiler is smart enough to reason with `N = Ndst - Nidx`, and the compilation is already specialized on `dst` and `idx` which univocally determine `N`. Anyways, it's an implementation detail not exposed to the user,  mostly a matter of conciseness, it's fine if you keep it as it is

Reply:
I commit a version for optimizing the process of generating colons.

```julia
julia> @code_typed ntuple(_->Colon(), Val(1))
CodeInfo(
1 ─     return (Colon(),)
) => Tuple{Colon}

julia> @code_typed ntuple(_->Colon(), Val(2))
CodeInfo(
1 ─     return (Colon(), Colon())
) => Tuple{Colon,Colon}

julia> @code_typed ntuple(_->Colon(), 3)
CodeInfo(
1 ── %1  = (n === 0)::Bool
└───       goto #3 if not %1
2 ──       goto #34
3 ── %4  = (n === 1)::Bool
└───       goto #5 if not %4
4 ──       goto #33
5 ── %7  = (n === 2)::Bool
└───       goto #7 if not %7
6 ──       goto #32
7 ── %10 = (n === 3)::Bool
└───       goto #9 if not %10
8 ──       goto #31
9 ── %13 = (n === 4)::Bool
└───       goto #11 if not %13
10 ─       goto #30
11 ─ %16 = (n === 5)::Bool
└───       goto #13 if not %16
12 ─       goto #29
13 ─ %19 = (n === 6)::Bool
└───       goto #15 if not %19
14 ─       goto #28
15 ─ %22 = (n === 7)::Bool
└───       goto #17 if not %22
16 ─       goto #27
17 ─ %25 = (n === 8)::Bool
└───       goto #19 if not %25
18 ─       goto #26
19 ─ %28 = (n === 9)::Bool
└───       goto #21 if not %28
20 ─       goto #25
21 ─ %31 = (n === 10)::Bool
└───       goto #23 if not %31
22 ─       goto #24
23 ─ %34 = invoke Base._ntuple(_2::Function, _3::Int64)::Tuple{Vararg{Colon,N} where N}
24 ┄ %35 = φ (#22 => (Colon(), Colon(), Colon(), Colon(), Colon(), Colon(), Colon(), Colon(), Colon(), Colon()), #23 => %34)::Tuple{Vararg{Colon,N} where N}
25 ┄ %36 = φ (#20 => (Colon(), Colon(), Colon(), Colon(), Colon(), Colon(), Colon(), Colon(), Colon()), #24 => %35)::Tuple{Vararg{Colon,N} where N}
26 ┄ %37 = φ (#18 => (Colon(), Colon(), Colon(), Colon(), Colon(), Colon(), Colon(), Colon()), #25 => %36)::Tuple{Vararg{Colon,N} where N}
27 ┄ %38 = φ (#16 => (Colon(), Colon(), Colon(), Colon(), Colon(), Colon(), Colon()), #26 => %37)::Tuple{Vararg{Colon,N} where N}
28 ┄ %39 = φ (#14 => (Colon(), Colon(), Colon(), Colon(), Colon(), Colon()), #27 => %38)::Tuple{Vararg{Colon,N} where N}
29 ┄ %40 = φ (#12 => (Colon(), Colon(), Colon(), Colon(), Colon()), #28 => %39)::Tuple{Vararg{Colon,N} where N}
30 ┄ %41 = φ (#10 => (Colon(), Colon(), Colon(), Colon()), #29 => %40)::Tuple{Vararg{Colon,N} where N}
31 ┄ %42 = φ (#8 => (Colon(), Colon(), Colon()), #30 => %41)::Tuple{Vararg{Colon,N} where N}
32 ┄ %43 = φ (#6 => (Colon(), Colon()), #31 => %42)::Tuple{Vararg{Colon,N} where N}
33 ┄ %44 = φ (#4 => (Colon(),), #32 => %43)::Tuple{Vararg{Colon,N} where N}
34 ┄ %45 = φ (#2 => (), #33 => %44)::Tuple{Vararg{Colon,N} where N}
└───       return %45
) => Tuple{Vararg{Colon,N} where N}
```

`ntuple(_->Colon(), Val(N))` returns directly, while `ntuple(_->Colon(), N)` encounter several branches.